src/components/Table/RowState.ts

```diff
@@ -0,0 +1,96 @@
+import { ObservableSet } from "mobx";
+import React, { MutableRefObject } from "react";
+import { GridDataRow } from "src/components/Table/GridTable";
+
+// Will be used in the next PR...
+// A parent row can be partially selected when some children are selected/some aren't.
+// export type SelectedState = "checked" | "unchecked" | "partial";
+
+/**
+ * Stores the collapsed & selected state of rows.
+ *
+ * I.e. this implements "collapse parent" --> "hides children", and
+ * "select parent" --> "select parent + children".
+ *
+ * There should be a single, stable `RowStateStore` instance per `GridTable`, so
+ * that children don't have to re-render even as we incrementally add/remove rows
+ * to the table (i.e. the top-level rows identity changes, but each row within it
+ * may not).
+ *
+ * We use mobx ObservableSets/ObservableMaps to drive granular re-rendering of rows/cells
+ * that need to change their toggle/select on/off in response to parent/child
+ * changes.
+ */
+export class RowState {
+  // A set of just row ids, i.e. not row.kind+row.id
+  readonly collapsedRows: ObservableSet<string>;
+  // Coming in future PR
+  // readonly selectedRows = new ObservableMap<string, "checked" | "unchecked" | "partial">();
+
+  /**
+   * Creates the `RowState` for a given `GridTable`.
+   */
+  constructor(private rows: MutableRefObject<GridDataRow<any>[]>, private persistCollapse: string | undefined) {
+    this.collapsedRows = new ObservableSet(persistCollapse ? readLocalCollapseState(persistCollapse) : []);
+  }
+
+  // Should be called in an Observer/useComputed to trigger re-renders
+  isCollapsed(id: string): boolean {
+    return this.collapsedRows.has(id) || this.collapsedRows.has("header");
+  }
+
+  toggleCollapsed(id: string): void {
+    const collapsedIds = [...this.collapsedRows.values()];
+
+    // We have different behavior when going from expand/collapse all.
+    if (id === "header") {
+      const isAllCollapsed = collapsedIds[0] === "header";
```

Comment:
This is the core of the old `useCollapsedIds` hook that does the same "what should/should not be collapsed now".